sui/tests/shared_objects_tests.rs

```diff
@@ -419,3 +425,77 @@ async fn shared_object_sync() {
         }
     }
 }
+
+#[tokio::test]
+async fn shared_object_on_gateway() {
+    let mut gas_objects = test_gas_objects();
+
+    // Get the authority configs and spawn them. Note that it is important to not drop
+    // the handles (or the authorities will stop).
+    let (configs, key_pairs) = test_authority_configs();
+    let _handles = spawn_test_authorities(gas_objects.clone(), &configs, &key_pairs).await;
+    let clients = create_authority_clients(&configs);
+    let path = tempfile::tempdir().unwrap().into_path();
+    let gateway = Arc::new(GatewayState::new_with_authorities(path, clients).unwrap());
+
+    // Publish the move package to all authorities and get the new package ref.
+    tokio::task::yield_now().await;
+    let package_ref = publish_counter_package(gas_objects.pop().unwrap(), &configs).await;
+
+    // Send a transaction to create a counter, but only to one authority.
+    tokio::task::yield_now().await;
+    let create_counter_transaction = move_transaction(
+        gas_objects.pop().unwrap(),
+        "Counter",
+        "create",
+        package_ref,
+        /* arguments */ Vec::default(),
+    );
+    let resp = gateway
+        .execute_transaction(create_counter_transaction)
+        .await
+        .unwrap();
+    let effects = resp.to_effect_response().unwrap().1;
+    let (shared_object_id, _, _) = effects.created[0].0;
+    // We need to have one gas object left for the final value check.
+    let last_gas_object = gas_objects.pop().unwrap();
+    let increment_amount = gas_objects.len();
+    let futures: Vec<_> = gas_objects
+        .into_iter()
+        .map(|gas_object| {
+            let g = gateway.clone();
+            let increment_counter_transaction = move_transaction(
+                gas_object,
+                "Counter",
+                "increment",
+                package_ref,
+                /* arguments */ vec![CallArg::SharedObject(shared_object_id)],
+            );
+            async move { g.execute_transaction(increment_counter_transaction).await }
+        })
+        .collect();
+
+    let replies: Vec<_> = futures::future::join_all(futures)
+        .await
+        .into_iter()
+        .collect();
+    assert_eq!(replies.len(), increment_amount);
```

Comment:
Are we sure this check always holds? The consensus may drop some transactions, so could it be the case the we get less replies than transactions we sent?

Reply:
hmm why and when would consensus drop some transactions?

Reply:
A narwhal advances to the next dag-round upon receiving a quorum of narwhal certificates from the previous round; narwhal then creates a new certificate (round r) by referring to a quorum of certificates from the previous round (round r-1). If `f` validators are consistently slower than the others, their narwhal certificates may never be referenced by other validators (they could consistently be ignored). Then, after 50 rounds the garbage collector kicks in and deletes those children-less certificates. Any narwhal transaction referenced by these certificates is thus lost.

Reply:
@lxfind @asonnino do we have consensus on this issue?

Reply:
For the purpose of this test, I don't think there is an issue. The gateway retries and the load is very low (~15 tx).